core/mention_api.php

```diff
@@ -0,0 +1,185 @@
+ * @return string The mentions tag.
+ */
+function mentions_tag() {
+	return config_get( 'mentions_tag', null, ALL_USERS, ALL_PROJECTS );
+}
+
+/**
+ * A method that takes in a text argument and extracts all candidate @ mentions
+ * from it.  The return list will not include the @ sign and will not include
+ * duplicates.  This method is mainly for testability and it doesn't take into
+ * consideration whether the @ mentions features is enabled or not.
+ *
+ * @param string $p_text The text to process.
+ * @return array of @ mentions without the @ sign.
+ * @private
+ */
+function mention_get_candidates( $p_text ) {
+	if( is_blank( $p_text ) ) {
+		return array();
+	}
+
+	$t_mentions_tag = mentions_tag();
+	preg_match_all( "/(?<!" . $t_mentions_tag. ")" . $t_mentions_tag . "[A-Za-z0-9_\.]+/", $p_text, $t_matches );
+
+	$t_mentions = array();
+
+	foreach( $t_matches[0] as $t_mention ) {
+		$t_mention = substr( $t_mention, 1 );
+
+		# "victor.boctor" is valid, "vboctor." should be "vboctor"
+		$t_mention = trim( $t_mention, '.' );
+		$t_mentions[$t_mention] = true;
+	}
+
+	$t_mentions = array_keys( $t_mentions );
+
+	return $t_mentions;
+}
+
+/**
+ * Given a string find the @ mentioned users.  The return list is a valid
+ * list of valid mentioned users.  The list will be empty if the mentions
+ * feature is disabled.
+ *
+ * @param string $p_text The text to process.
+ * @return Array with valid usernames as keys and their ids as values.
+ */
+function mention_get_users( $p_text ) {
+	if ( !mention_enabled() ) {
+		return array();
+	}
+
+	$t_matches = mention_get_candidates( $p_text );
+	if( empty( $t_matches )) {
+		return array();
+	}
+
+	$t_mentioned_users = array();
+
+	foreach( $t_matches as $t_candidate ) {
+		if( $t_user_id = user_get_id_by_name( $t_candidate ) ) {
+			if( false === $t_user_id ) {
+				continue;
+			}
+
+			$t_mentioned_users[$t_candidate] = $t_user_id;
+		}
+	}
+
+	return $t_mentioned_users;
+}
+
+/**
+ * Process users that are mentioned on the specified bug.
+ *
+ * @param int $p_bug_id The bug id.
+ * @param array $p_mentioned_user_ids An array of user ids
+ * @param string $p_message The message containing the mentions.
+ * @return void
+ */
+function mention_process_user_mentions( $p_bug_id, $p_mentioned_user_ids, $p_message = '' ) {
+	email_user_mention( $p_bug_id, $p_mentioned_user_ids, $p_message );
+}
+
+/**
+ * Format and hyperlink mentions
+ *
+ * @param string $p_text The text to process.
+ * @param bool $p_html true for html, false otherwise.
+ * @return string The processed text.
+ */
+function mention_format_text( $p_text, $p_html = true ) {
+	if ( !mention_enabled() ) {
+		return $p_text;
+	}
+
+	$t_text = $p_text;
+
+	$t_mentioned_users = mention_get_users( $p_text );
+	if( empty( $t_mentioned_users ) ) {
+		return $p_text;
+	}
+
+	$t_mentions_tag = mentions_tag();
+	$t_formatted_mentions = array();
+
+	foreach( $t_mentioned_users as $t_username => $t_user_id  ) {
+		if( user_exists( $t_user_id ) ) {
```

Comment:
Removed.  mention_get_users() will never return non-existent users, but in the model where we had placeholders, this was a possible case.